components/nexus-repository/src/main/java/org/sonatype/nexus/repository/view/Content.java

```diff
@@ -100,4 +112,85 @@ public String toString() {
+   */
+  @VisibleForTesting
+  static final String P_ETAG = "etag";
+
+  /**
+   * Extracts non-format specific content attributes into the passed in {@link AttributesMap} (usually originating from
+   * {@link Content#getAttributes()}) from passed in {@link Asset} and format required hashes.
+   */
+  public static void extractFromAsset(final Asset asset,
+                                      final Iterable<HashAlgorithm> hashAlgorithms,
+                                      final AttributesMap contentAttributes)
+  {
+    checkNotNull(asset);
+    checkNotNull(hashAlgorithms);
+    final NestedAttributesMap assetAttributes = asset.attributes().child(P_CONTENT_ATTRIBUTES);
+    final DateTime lastModified = toDateTime(assetAttributes.get(P_LAST_MODIFIED, Date.class));
+    final String etag = assetAttributes.get(P_ETAG, String.class);
+
+    final NestedAttributesMap checksumAttributes = asset.attributes().child(StorageFacet.P_CHECKSUM);
+    final Map<HashAlgorithm, HashCode> hashCodes = Maps.newHashMap();
+    for (HashAlgorithm algorithm : hashAlgorithms) {
+      final HashCode hashCode = HashCode.fromString(checksumAttributes.require(algorithm.name(), String.class));
+      hashCodes.put(algorithm, hashCode);
+    }
+
+    contentAttributes.set(Content.CONTENT_LAST_MODIFIED, lastModified);
+    contentAttributes.set(Content.CONTENT_ETAG, etag);
+    contentAttributes.set(Content.CONTENT_HASH_CODES_MAP, hashCodes);
+    contentAttributes.set(CacheInfo.class, CacheInfo.extractFromAsset(asset));
+  }
+
+  /**
+   * Applies non-format specific content attributes onto passed in {@link Asset} from passed in {@link AttributesMap}
+   * (usually originating from {@link Content#getAttributes()}).
+   */
+  public static void applyToAsset(final Asset asset, final AttributesMap contentAttributes) {
+    checkNotNull(asset);
+    checkNotNull(contentAttributes);
+    final NestedAttributesMap assetAttributes = asset.attributes().child(P_CONTENT_ATTRIBUTES);
+    assetAttributes.set(P_LAST_MODIFIED, toDate(contentAttributes.get(Content.CONTENT_LAST_MODIFIED, DateTime.class)));
+    assetAttributes.set(P_ETAG, contentAttributes.get(Content.CONTENT_ETAG, String.class));
+    final CacheInfo cacheInfo = contentAttributes.get(CacheInfo.class);
+    if (cacheInfo != null) {
+      CacheInfo.applyToAsset(asset, cacheInfo);
+    }
+  }
+
+  /**
+   * Maintains the "last modified" attribute of the content by setting it to "now". It accepts {@code null}s, and the
+   * returned instance will have changes applied. Never returns {@code null}.
+   */
+  @Nonnull
+  public static AttributesMap maintainLastModified(final Asset asset,
+                                                   @Nullable final AttributesMap contentAttributes)
+  {
+    AttributesMap ca = contentAttributes;
+    if (ca == null) {
+      ca = new AttributesMap();
+    }
+    if (!ca.contains(Content.CONTENT_LAST_MODIFIED)) {
```

Comment:
Is this enough? What happens if the blob attached to the asset changes, in that case the attribute map will already contain a last-modified date and won't be updated. See https://issues.sonatype.org/browse/NEXUS-8937